### PR TITLE
Add HTTP Provider

### LINKED ISSUE #1053: Feature request: HTTP backend
Hey,

This feature is called [Service Discovery](https://tyk.io/docs/tyk-api-gateway-v-2-0/api-management/service-discovery/) in [Tyk](https://tyk.io). The basic idea is that there's external API I can setup that Traefik uses to "resolve" API endpoints. For example I could setup an application and store all routes in database, then expose an endpoint that would return these routes, so Traefik can query and use them to route incoming requests. The flow is as follows:

1. Incoming request arrives to Traefik for some domain/port/path(optionally)
2. Traefik asks HTTP backend to resolve incoming query
3. Traefik uses received information to properly route incoming request
4. Next incoming request arrives for the same domain/port/path(optionally)
5. Traefik uses cached information in 2. to properly route this request
6. Traefik purges cache after 10+ seconds and will query HTTP backend next time such request arrives

What do you think?

### PR DESCRIPTION
**Pull request by dtomcej**
### What does this PR do?

This PR introduces a new provider (http) that polls an endpoint for a published configuration, and provides it to the server for use.

### Motivation

Fixes #1053

### More

- [x] Added/updated tests
- [x] Added/updated documentation

### CODE REVIEW

File: docs/content/providers/overview.md [hunk 1]
```diff
@@ -35,9 +35,10 @@ Below is the list of the currently supported providers in Traefik.
 | [Rancher](./rancher.md)               | Orchestrator | Label                      |
 | [File](./file.md)                     | Manual       | TOML/YAML format           |
 | [Consul](./consul.md)                 | KV           | KV                         |
-| [etcd](./etcd.md)                     | KV           | KV                         |
+| [Etcd](./etcd.md)                     | KV           | KV                         |
 | [Redis](./redis.md)                   | KV           | KV                         |
 | [ZooKeeper](./zookeeper.md)           | KV           | KV                         |
+| [Http](./http.md)                     | Manual       | JSON format                |
```
Comment (notsureifkevin) [inline docs/content/providers/overview.md:41]: https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol

NIT: I noticed we use `Http` and `HTTP` interchangeably throughout the documentation and I recommend the naming standard as noted in the RFC should be used for consistency.

File: docs/content/providers/overview.md [hunk 2]
```diff
@@ -35,9 +35,10 @@ Below is the list of the currently supported providers in Traefik.
 | [Rancher](./rancher.md)               | Orchestrator | Label                      |
 | [File](./file.md)                     | Manual       | TOML/YAML format           |
 | [Consul](./consul.md)                 | KV           | KV                         |
-| [etcd](./etcd.md)                     | KV           | KV                         |
+| [Etcd](./etcd.md)                     | KV           | KV                         |
```
Comment (notsureifkevin) [inline docs/content/providers/overview.md:38]: while `etcd` has no RFC, it is commonly referred to as `etcd` by the publishers of the project. You can also find examples all over the web where the `e` is capitalized, however, I noticed we use `Etcd` consistently throughout the documentation so this makes sense.

**Comment by notsureifkevin** [comment]
If an operator wants to utilize this with a `https` endpoint that doesn't have a valid cert, for any number of reasons, I'm assuming this would fail.

Should we introduce a flag which allows bypassing a cert check?

File: docs/content/providers/overview.md (hunk 2 above)
Comment (dtomcej) [reply docs/content/providers/overview.md:38]: Also when used as a proper noun, usually capitalized.

**Comment by dtomcej** [comment]
Closing in favour of https://github.com/containous/traefik/pull/6976

**Pull request closed without merging**